pex/resolve/pep_691/fingerprint_service.py

```diff
@@ -0,0 +1,217 @@
+                # to an even lower chunk size to be safe: https://www.sqlite.org/limits.html
+                chunk_size = 100
+                for index in range(0, len(urls), chunk_size):
+                    chunk = urls[index : index + chunk_size]
+                    with closing(
+                        conn.execute(
+                            "SELECT url, algorithm, hash FROM hashes WHERE url IN ({})".format(
+                                ", ".join(repeat("?", len(chunk)))
+                            ),
+                            tuple(chunk),
+                        )
+                    ) as cursor:
+                        for url, algorithm, hash_ in cursor:
+                            yield _FingerprintedURL(
+                                url=url, fingerprint=Fingerprint(algorithm=algorithm, hash=hash_)
+                            )
+
+    def _cache(self, fingerprinted_urls):
+        # type: (Sequence[_FingerprintedURL]) -> None
+
+        with TRACER.timed(
+            "Caching {count} fingerprints in database".format(count=len(fingerprinted_urls))
+        ):
+            with sqlite3.connect(self._path) as conn:
+                conn.executemany(
+                    "INSERT OR REPLACE INTO hashes (url, algorithm, hash) VALUES (?, ?, ?)",
+                    tuple(
+                        (
+                            fingerprinted_url.url,
+                            fingerprinted_url.fingerprint.algorithm,
+                            fingerprinted_url.fingerprint.hash,
+                        )
+                        for fingerprinted_url in fingerprinted_urls
+                    ),
+                ).close()
+
+    def _safe_request(self, endpoint):
+        # type: (Endpoint) -> Union[Project, Error]
+        return catch(self._api.request, endpoint)
+
+    def fingerprint(
+        self,
+        endpoints,  # type: Set[Endpoint]
+        artifacts,  # type: Iterable[PartialArtifact]
+    ):
+        # type: (...) -> Iterator[PartialArtifact]
+        """Attempts to fingerprint all artifacts that are missing a fingerprint.
+
+        Fingerprints are obtained via the PEP-691 JSON API and are not verified.
+
+        :param artifacts: The artifacts to fill in missing fingerprints for.
+        :return: An iterator over all the artifacts given, where some returned artifacts may have
+                 previously missing fingerprints filled in (but not verified).
+        """
+        # N.B.: In the absence of having an advertised fingerprint for an artifact, Pex will
+        # download the artifact and hash it at higher layers; so simply warning about the error is
+        # enough; we need not fail the whole endeavor. The partial artifact just remains partial
+        # and becomes more expensive to lock. The warning is likely un-actionable by the user but
+        # also rare; so it can be reported to maintainers in case the 5s default sqlite3 timeout is
```

Comment:
Would it make sense to encourage users to post issues to pex repo when they encounter this?

Reply:
No, but yes if they encounter it often. I'll add to the warning with more verbage.